4단계 - 요금 조회

ssamzag opened the pull request:
안녕하세요 오경태 리뷰어님!!
일요일 밤 잘 보내고 계신지요? 벌써 월요일이라니.. 
주말은 정말 시간이 빠르게 가는 것 같아요.. 😢 

이번 미션에서 가장 고민했던 것 중 하나가 요금관련 클래스를 어느 패키지에 넣느냐였는데요. 우선은 Sections가 요금을 의존하고 있어서 같은 패키지인 line domain에 묶어 뒀습니다. 하지만 경로찾기와 요금이 비즈니스적으로 묶여 있기 때문에 path 에 넣어주는 게 맞는 거 같기도 하구요.
사소한 것도 참 어렵네요;; 

이번 리뷰도 잘 부탁드립니다!!

ssamzag committed:
feat: 4단계 미션 - 요금조회 구현

ssamzag committed:
style: 파라미터명 s -> n 통일

ssamzag committed:
refactor: 거리기준 계산 enum화

ssamzag committed:
chore: 미사용 메서드 삭제

ssamzag committed:
refactor: 파라미터명 n -> age 변경

ssamzag committed:
refactor: 거리요금계산 호출메서드 Distance 인스턴스로 이동

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -78,22 +84,38 @@ void findPath() {
                 () -> assertThat(pathResponse.getStations().get(3).getId()).isEqualTo(강남역.getId()),
-                () -> assertThat(pathResponse.getDistance()).isEqualTo(18));
+                () -> assertThat(pathResponse.getDistance()).isEqualTo(18),
+                () -> assertThat(pathResponse.getFare()).isEqualTo(1450));
```
현재 구조라면, 교대역에서 강남역으로 가는 최단 경로를 거치는 노선들은 3호선, 신분당선인 것 같아요..! 
신분당선에 노선 추가 요금이 1,000원이 적용되어있는데 추가 요금이 붙지 않은 것 같아요 :) 
기본요금 1,250 + 5km 추가 금액 200 + 신분당선 추가 요금 1000 = 2,450)
한번 확인해주실 수 있으실까요?

ssamzag replied:
앗.. 확인해서 수정해 놓겠습니다!!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Discount.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Arrays;
+import java.util.function.Predicate;
+
+public enum Discount {
```
어떤 할인인지에 대한 구체적인 네이밍이면 더 좋을 것 같아요 :)

ssamzag replied:
넵~ 적당한 네이밍을 붙여야겠네요

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Discount.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final int discountRate;
+    private final int discount;
+    private final Predicate<Integer> ageMatchingExp;
```
👍 👍

ohtaeg commented on src/main/java/nextstep/subway/line/domain/DistanceType.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public enum DistanceType {
+    BASIC(distance -> distance <= 10, distance -> 0),
+    MIDDLE(distance -> distance > 10 && distance <= 50, distance -> calculate(distance, 10, 5)),
+    LONG(distance -> distance > 50, distance -> calculate(distance, 50, 8) + MIDDLE.calculateDistanceFare(50));
+
+    private final Predicate<Integer> distanceMatchingExp;
+    private final Function<Integer, Integer> distanceCalculateFunc;
```
거리별 요금 정책을 enum으로 잘 표현해주셨어요 👍
사소하지만 distanceCalculateFunc에 쓰이는 숫자들을 상수로 치환하여 어떤 숫자들인지 의미를 부여하도록 하면
더 좋을 것 같아요 :)

ssamzag replied:
넵 원시값을 그대로 써서 찝찝했네요.. 수정해 놓겠습니다!!

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -200,6 +202,10 @@ private Section getSection(List<Section> sections, Station station, Station next
                 .orElseThrow(() -> new IllegalArgumentException("역이 존재하지 않습니다"));
     }
 
+    public Fare fare(Integer age) {
+        return Fare.calculate(totalDistance(), age, maxSurcharge());
+    }
```
이 부분은 같이 고민해보면 좋을 것 같아요 :)
Sections는 Fare를 의존하고 있는데요! 개인적으로 Sections가 Fare를 의존할 필요가 있는지 고민이 되는 부분이네요..!
PathService에서 최단 경로 찾기를 통해 Sections를 도출한 상태에서 
Sections를 통해 요금을 계산하도록 메시지를 보내고 있는데요 :)

Sections가 요금을 계산하기 위해 제공하는 정보는 **노선에 추가 요금**과 **총 거리** 인 것 같아요 :)
그렇다면 Sections에서 꼭 Fare를 알고있지 않더라도 외부에 **노선에 추가 요금**과 **총 거리** 만 제공해준다면
요금을 조회할 수 있을 것 같아요! 이미 Fare 객체가 요금을 계산하도록 정보를 다음과 같이 받고 있기도 하구요 :)
```
public static Fare calculate(Distance distance, Integer age, int surcharge) {
```



만약 각 도메인 영역들이 모듈로 분리된 환경으로 생각해보았을 때 계산하는 책임을 Path 도메인 뿐만 아니라 Line까지 전파되는 부분이라 Path 패키지 내에 한정으로 계산하는 책임을 가질 수 있도록 할 수 있을 것 같기도 하네요!
그렇게 되면 주셨던 질문도 자연스럽게 요금 관련 클래스들이 path 패키지에 위치하게 될 것 같구요 😄 

> 이번 미션에서 가장 고민했던 것 중 하나가 요금관련 클래스를 어느 패키지에 넣느냐였는데요. 우선은 Sections가 요금을 의존하고 있어서 같은 패키지인 line domain에 묶어 뒀습니다. 하지만 경로찾기와 요금이 비즈니스적으로 묶여 있기 때문에 path 에 넣어주는 게 맞는 거 같기도 하구요. 사소한 것도 참 어렵네요;;

물론 요금을 계산하기 위한 정보를 잘 알고 있는 객체는 Sections이기도 하여 Sections를 통해 메시지를 보내는 설계도
이해가 가는 부분이라 고민이네요 :)

물론 위 내용이 틀릴수도 있고, 민호님 생각과 다를수도 있기에 조심스러운 부분입니다 🙏 
어느 부분이 좋을지 트레이드 오프하면 될 것 같은데, ~~호민~~민호님은 어떻게 생각하시나요 😄

ssamzag replied:
말씀하신대로 Sections에서 Fare 계산 위임을 빼는 게 의도한대로 path 패키지에 위치시킬 수도 있겠네요. Sections의 역할이 많아지는 것도 좋아 보이지 않구요. 괜한 욕심을 부린 것 같네요.. 😅 
path 서비스에서 fare를 그대로 호출하도록 수정하겠습니다. 
감사합니다! 👍

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -57,6 +70,10 @@ public List<Section> getSections() {
         return sections.getSections();
     }
 
+    public Integer getSurcharge() {
```
primitive 타입이 아닌 Wrapper 클래스로 반환해주시는 의도가 따로 있으실까요 :) ?

ssamzag replied:
surcharge 필드가 Integer 라서 get도 동일하게 Integer로 반환하도록 했는데 생각해보니 wrapper 클래스를 사용하면 안 되겠네요. id 컬럼은 0이 되면 안되기에 래퍼 클래스를 썼지만, 저 필드는 null을 허용 안 하는 게 맞는 것 같습니다. 
수정하도록 할게요!

ohtaeg reviewed (changes requested):
4단계 미션도 구현을 잘해주셨어요 👍 👍 
소소한 코멘트를 드렸는데 마지막으로 고민해보면 좋을 포인트가 있기에 Request chagnes 하였습니다 :)
확인 후 재요청 주시면 감사하겠습니다 😄 
꿀같은 주말이 벌써 끝나버렸네요 😭 
다시 주말을 기다리며.. 이번주도 화이팅입니다 🔥

ssamzag committed:
refactor&style: surcharge 저장 및 Discount enum 네이밍 변경
- Discount -> AgeDiscountType
- Line 엔티티 save 시 surcharge 저장
- Fare 관련 클래스 Path 패키지로 이동

ssamzag committed:
refactor: DistanceType 원시값 의미있는 상수값으로 변경

ssamzag commented:
안녕하세요. ~~경태~~ 태경 리뷰어님! 😄 
4단계 피드백 반영분 리뷰 요청드립니다. 
월요일 새벽에 리뷰 올라와서 엄청 놀랬습니다. 강철 체력이신지요!! 😯😯
피곤하실텐데 컨디션 관리 잘 하시고 야근 없는 무난한 한 주를 보내시길 바라겠습니다!! 🔥🔥

ssamzag committed:
refactor: 필드 및 원시값 네이밍 의미있게 변경

ssamzag committed:
docs: 4단계 미션 문서정리

ssamzag committed:
docs: 문서수정

ohtaeg commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -19,26 +20,27 @@ public class Line extends BaseEntity {
     @ColumnDefault("0")
-    private Integer surcharge;
+    @Column(nullable = false)
+    private int surcharge;
```
피드백 반영 👍

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -1,7 +1,7 @@
 
 import nextstep.subway.line.application.LineService;
-import nextstep.subway.line.domain.Fare;
+import nextstep.subway.path.domain.Fare;
```
👍 👍

ohtaeg commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -31,7 +31,7 @@ public PathResponse findShortestDistancePath(Integer age, Long sourceId, Long ta
         List<Station> shortestStations = PathFinder.find(new Dijkstra(source, target, sections));
         Sections filteredSections = sections.filteredBy(shortestStations);
-        Fare fare = filteredSections.fare(age);
+        Fare fare = Fare.calculate(filteredSections.totalDistance(), age, filteredSections.maxSurcharge());
```
의견이 다를수도 있는 부분이였기에 조심스러운 피드백이였는데 반영 해주셔서 감사드립니다 😄 🙏

ssamzag replied:
좋은 건 반영해야쥬 😄

ohtaeg commented on src/main/java/nextstep/subway/path/domain/DistanceType.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public enum DistanceType {
```
이전 Discount는 거리에 대한 타입이군요 👍

ohtaeg commented on src/test/java/nextstep/subway/path/PathAcceptanceTest.java:
```diff
@@ -85,7 +85,7 @@ void findPathWithNoMember() {
+                () -> assertThat(pathResponse.getFare()).isEqualTo(2450));
     }
 
     /**
```
해당 테스트가 에러가 나는 것 같아요 😭 
아마 리팩토링하시면서 해당 테스트 코드를 따로 변경이 되지 않은 것 같아요 :)
로직 검증을 해보니 1,250 + 노선 추가비용 1,000원을 더해서 2,250원인 상태에서
거리가 15km를 넘어 200원 추가되어 2,450원이 되고
2,450원에서 청소년 350원을 공제한 2,100원에서 20%만 할인을 적용하면
1,680원이 계산이 되네요 :) 실제 테스트 결과 값도 일치하구요!
880원 부분만 수정해주시면 좋을 것 같아요 😄

ssamzag replied:
커밋전에 전체 테스트를 한 번 돌렸어야 했는데 하나만 한 결과가.. 😨 바로 수정했네요.

ohtaeg commented:
> 월요일 새벽에 리뷰 올라와서 엄청 놀랬습니다. 강철 체력이신지요!! 😯😯
피곤하실텐데 컨디션 관리 잘 하시고 야근 없는 무난한 한 주를 보내시길 바라겠습니다!! 🔥🔥

ohtaeg commented:
> 월요일 새벽에 리뷰 올라와서 엄청 놀랬습니다. 강철 체력이신지요!! 😯😯
피곤하실텐데 컨디션 관리 잘 하시고 야근 없는 무난한 한 주를 보내시길 바라겠습니다!! 🔥🔥

쓰러지기 일보 직전에 위해주는 말씀 덕분에 힘이 나네요 💪 력 💪
민호님도 미션하시느라 많이 힘드실텐데 남은 기간동안 컨디션 조절 잘하시구 이번주도 무난한 한 주되길 기도하겠습니다 🙏

ohtaeg reviewed (approved):
마지막까지 피드백을 잘반영해주셨어요 👍 👍 
마지막으로 사소한 코멘트 하나를 드렸는데 반영해주시고 마무리해주시면 감사하겠습니다 :)
잘해주셔서 이만 마무리해도 될 것 같아요!!

ATDD + TDD 미션 완주를 축하드립니다. 🎉 
부족하지만 끝까지 따라와주셔서, 그리고 피드백 반영을 잘해주셔서 감사드립니다 🙏 
리뷰하면서 마치 대화하는 것처럼 즐겁게 리뷰할 수 있었습니다 💪 😄 
남은 미션도 화이팅하세요!! 고생하셨습니다 🙇
